tests/blockchain/eth/contracts/main/mining_adjudicator/test_mining_adjudicator.py

```diff
@@ -293,21 +304,69 @@ def test_evaluate_cfrag(testerchain, escrow, adjudicator_contract):
     wrong_args = args[:]
-    wrong_args[7] = signed_miner_umbral_public_key
+    wrong_args[7] = signed_wrong_miner_umbral_public_key
     with pytest.raises((TransactionFailed, ValueError)):
         tx = adjudicator_contract.functions.evaluateCFrag(*wrong_args).transact()
         testerchain.wait_for_receipt(tx)
 
     # Initial arguments were correct
     assert not adjudicator_contract.functions.evaluatedCFrags(data_hash).call()
-    tx = adjudicator_contract.functions.evaluateCFrag(*args).transact()
+    tx = adjudicator_contract.functions.evaluateCFrag(*args).transact({'from': investigator})
     testerchain.wait_for_receipt(tx)
     assert adjudicator_contract.functions.evaluatedCFrags(data_hash).call()
+    # Penalty was increased because it's the second violation
     assert 790 == escrow.functions.minerInfo(miner).call()[0]
-    # TODO tests for penalty/reward calculation
-    # TODO tests for events
+    assert 105 == escrow.functions.rewardInfo(investigator).call()
+
+    events = evaluation_log.get_all_entries()
+    assert 3 == len(events)
+    event_args = events[2]['args']
+    assert data_hash == event_args['evaluationHash']
+    assert miner == event_args['miner']
+    assert investigator == event_args['investigator']
+    assert not event_args['correctness']
+
+    # Prepare corrupted data again
+    capsule, cfrag = fragments(metadata)
+    capsule_bytes = capsule.to_bytes()
+    # Corrupt proof
+    cfrag.proof.bn_sig = CurveBN.gen_rand(capsule.params.curve)
+    cfrag_bytes = cfrag.to_bytes()
+    hash_ctx = hashes.Hash(hashes.SHA256(), backend=backend)
+    hash_ctx.update(capsule_bytes + cfrag_bytes)
+    data_hash = hash_ctx.finalize()
+    capsule_signature_by_requester = sign_data(capsule_bytes, requester_umbral_private_key)
+    capsule_signature_by_requester_and_miner = sign_data(capsule_signature_by_requester, miner_umbral_private_key)
+    cfrag_signature_by_miner = sign_data(cfrag_bytes, miner_umbral_private_key)
+    evidence = IndisputableEvidence(capsule, cfrag, ursula=None)
+    evidence_data = evidence.precompute_values()
+    args = (capsule_bytes,
+            capsule_signature_by_requester,
+            capsule_signature_by_requester_and_miner,
+            cfrag_bytes,
+            cfrag_signature_by_miner,
+            requester_umbral_public_key_bytes,
+            miner_umbral_public_key_bytes,
+            signed_miner_umbral_public_key,
+            evidence_data)
+
+    assert not adjudicator_contract.functions.evaluatedCFrags(data_hash).call()
+    tx = adjudicator_contract.functions.evaluateCFrag(*args).transact({'from': investigator})
+    testerchain.wait_for_receipt(tx)
+    assert adjudicator_contract.functions.evaluatedCFrags(data_hash).call()
+    # Penalty was decreased because it's more than maximum available percentage of value
+    assert 692 == escrow.functions.minerInfo(miner).call()[0]
+    assert 154 == escrow.functions.rewardInfo(investigator).call()
```

Comment:
I think it could be good to show how these numbers (692, 154) were computed. The same applies to more round stake amounts, but I think here's more noticeable.

Reply:
✔️